gateway-service/src/main/java/org/java/mentorship/gateway/exception/handler/GeneralExceptionHandler.java

```diff
@@ -1,31 +1,68 @@
 package org.java.mentorship.gateway.exception.handler;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import feign.FeignException;
 import jakarta.servlet.http.HttpServletRequest;
+import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
-import org.java.mentorship.gateway.exception.domain.APIErrorResponse;
+import org.java.mentorship.contracts.common.dto.ErrorResponse;
+import org.java.mentorship.gateway.exception.domain.GatewayErrorResponse;
 import org.java.mentorship.gateway.exception.domain.GatewayException;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.ExceptionHandler;
 import org.springframework.web.bind.annotation.RestControllerAdvice;
 
+import java.nio.charset.StandardCharsets;
+
 @RestControllerAdvice
 @Slf4j
+@RequiredArgsConstructor
 public class GeneralExceptionHandler {
 
+    private final ObjectMapper objectMapper;
+
     @ExceptionHandler(GatewayException.class)
-    public ResponseEntity<APIErrorResponse> handle(final GatewayException exception, final HttpServletRequest request) {
+    public ResponseEntity<ErrorResponse> handleGatewayException(final GatewayException exception) {
+        if (exception.getSourceException().isPresent()) {
+            log.error(String.format("Caught exception: %s (%s)",
+                    exception.getSourceException().get().getMessage(),
+                    exception.getSourceException().get().getClass()));
+        }
         return ResponseEntity.status(exception.getStatusCode())
-                .body(new APIErrorResponse(exception.getMessage()));
+                .body(exception.getErrorResponse());
+    }
+
+    @ExceptionHandler(FeignException.class)
+    public ResponseEntity<ErrorResponse> handleFeignException(final FeignException exception, final HttpServletRequest request) {
+        ErrorResponse errorResponse = new GatewayErrorResponse("No message");
+
+        if (exception.responseBody().isPresent()) {
+            String responseString = StandardCharsets.UTF_8.decode(exception.responseBody().get()).toString();
+            try {
+                errorResponse = objectMapper.readValue(responseString, ErrorResponse.class);
+            } catch (JsonProcessingException e) {
+                log.error(String.format("[%s] Caught exception when parsing JSON: %s (%s)",
```

Comment:
In order to have the full stack trace, I'd use this method as this:

log.error(String.format("[%s] Caught exception when parsing JSON", request.getRequestURI()), e)

This would print the error class, message and stack trace as well